services/src/main/java/io/grpc/services/BinaryLog.java

```diff
@@ -275,13 +282,38 @@ static BinaryLog getLog(String fullMethodName) {
     return DEFAULT_FACTORY.getLog(fullMethodName);
   }
 
+  static byte[] getCallIdForServer(Context context) {
+    CallId callId = BinaryLogProvider.SERVER_CALL_ID_CONTEXT_KEY.get(context);
+    if (callId == null) {
+      return emptyCallId;
+    }
+    return callId.id;
+  }
+
+  static byte[] getCallIdForClient(CallOptions callOptions) {
+    CallId callId = callOptions.getOption(BinaryLogProvider.CLIENT_CALL_ID_CALLOPTION_KEY);
+    if (callId == null) {
+      return emptyCallId;
+    }
+    return callId.id;
+  }
+
+  static SocketAddress getPeerSocket(Attributes streamAttributes) {
+    SocketAddress peer = streamAttributes.get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR);
+    if (peer == null) {
+      return UNKNOWN_SOCKET;
+    }
+    return peer;
+  }
+
   @Override
   public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
       final MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    final byte[] callId = getCallIdForClient(callOptions);
     return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
       @Override
       public void start(Listener<RespT> responseListener, Metadata headers) {
-        writer.logSendInitialMetadata(headers, CLIENT, dumyCallId, DUMMY_SOCKET);
+        writer.logSendInitialMetadata(headers, CLIENT, callId);
```

Comment:
Since you already have a `CallId` type, why not change the writer interface to accept that instead of `byte[]`?

Reply:
Done. It was a good suggestion because now we can encapsulate two byte[] of size 8.

Reply:
Maybe make `CallId` to keep two `long`s? Then the `CallId` can be completely immutable.